feat: support for coded phy packet and power level

Louka-Yerly opened the pull request:
Unlike the BLE 1M and BLE 2M packets, CODED PHY packets contain two entries in results/<sim_id>/*.Tx.csv files (one for FEC1 and a second for FEC2). This modification takes this into account when creating a BLE entry for Wireshark.
This modification also adds the power of transmissions for Wireshark.

Commit:
feat: support for coded phy packet and power level

Unlike the BLE 1M and BLE 2M packets, CODED PHY packets contain
two entries in results/<sim_id>/*.Tx.csv files (one for FEC1 and a
second for FEC2). This modification takes this into account when
creating a BLE entry for Wireshark.
This modification also adds the power of transmissions for Wireshark.

Signed-off-by: Louka Yerly <louka.yerly@hes-so.ch>

aescolar reviewed (changes requested):
Thanks @Louka-Yerly , wouldn't we need to do something like this:
https://github.com/aescolar/ext_2G4_phy_v1/commit/fb054b5598a1b742a7e98de5d98ccf02dd2e62a7
So we ensure packets are in order in the output? (another not-coded Tx may start between the FEC1 and FEC2 and get in the trace before the coded packet)

Louka-Yerly commented:
Thank you for this suggestion for improvement @aescolar, it's a very good comment.

The code suggestion you made should work normally, but for me it generates an error.

As I have seen, a FEC1 is not necessarily followed by a FEC2. The extract below shows part of the file `d_2G4_00.Tx.csv`. As you can see, on the penultimate packet, there is a FEC1 but no corresponding FEC2.

<pre>
start_time,end_time,center_freq,phy_address,modulation,power_level,abort_time,recheck_time,packet_size,packet
20050763,20051138,2.000000,0x8E89BED6,80,0.000000,18446744073709551615,20051453,1,00
20051139,20053658,2.000000,0x00000000,80,0.000000,18446744073709551615,20053687,39,42 22 14 45 77 04 F5 DB 1B 2A 3D E7 0D 19 2B FD 57 48 4C EF 1D 27 17 68 10 CA FD E3 19 CF 2C 58 87 63 28 8F 02 86 D8
20053688,20054063,26.000000,0x8E89BED6,80,0.000000,18446744073709551615,20074585,1,00
20054064,20056583,26.000000,0x00000000,80,0.000000,18446744073709551615,20056612,39,42 22 14 45 77 04 F5 DB 1B 2A 3D E7 0D 19 2B FD 57 48 4C EF 1D 27 17 68 10 CA FD E3 19 CF 2C 58 87 63 28 8F 02 86 D8
20056613,20056988,80.000000,0x8E89BED6,80,0.000000,18446744073709551615,20074585,1,00
20056989,20059508,80.000000,0x00000000,80,0.000000,18446744073709551615,20060181,39,42 22 14 45 77 04 F5 DB 1B 2A 3D E7 0D 19 2B FD 57 48 4C EF 1D 27 17 68 10 CA FD E3 19 CF 2C 58 87 63 28 8F 02 86 D8
20076856,20077231,2.000000,0x8E89BED6,80,0.000000,18446744073709551615,20081452,1,00
20077232,20079751,2.000000,0x00000000,80,0.000000,18446744073709551615,20079780,39,42 22 14 45 77 04 F5 DB 1B 2A 3D E7 0D 19 2B FD 57 48 4C EF 1D 27 17 68 10 CA FD E3 19 CF 2C 58 87 63 28 8F 02 86 D8
20079781,20080156,26.000000,0x8E89BED6,80,0.000000,18446744073709551615,20081452,1,00
20080157,20082676,26.000000,0x00000000,80,0.000000,18446744073709551615,20082705,39,42 22 14 45 77 04 F5 DB 1B 2A 3D E7 0D 19 2B FD 57 48 4C EF 1D 27 17 68 10 CA FD E3 19 CF 2C 58 87 63 28 8F 02 86 D8
<b>20082706,20083081,80.000000,0x8E89BED6,80,0.000000,20082947,20082957,1,00</b>
20163098,20163473,2.000000,0x8E89BED6,80,0.000000,18446744073709551615,20168894,1,00
20163474,20165993,2.000000,0x00000000,80,0.000000,18446744073709551615,20166022,39,42 22 14 45 77 04 F5 DB 1B 2A 3D E7 0D 19 2B FD 57 48 4C EF 1D 27 17 68 10 CA FD E3 19 CF 2C 58 87 63 28 8F 02 86 D8
</pre>

I notice that the "abort_time" field is quite different from the other packets, but I don't know what that means.

aescolar commented:
@Louka-Yerly
> As I have seen, a FEC1 is not necessarily followed by a FEC2. 

You are indeed right. A coded phy packet Tx may be aborted before the FEC2 starts.
https://github.com/aescolar/ext_2G4_phy_v1/commit/f2ea90ff17699de398b911dbd652833900575fba

```diff --git a/dump_post_process/csv2pcap b/dump_post_process/csv2pcap
index e90f282..aa5ecde 100755
--- a/dump_post_process/csv2pcap
+++ b/dump_post_process/csv2pcap
@@ -39,6 +39,15 @@ def get_data_and_len(row):
 
        return (orig_len, pdu_crc)
 
+def is_coded_FEC2(row):
+       if not row:
+               return False
+       if int(row['modulation'], 10) != P2G4_MOD_BLE_CODED:
+               return False
+       if int(row['packet_size'], 10) < 5: # It must at least have a header and CRC (2+3 bytes)
+               return False
+       return True
+
 def write(outfile, *inputs, snaplen = SNAPLEN, basetime=None):
        #For information on the pcap format see https://wiki.wireshark.org/Development/LibpcapFileFormat
 
@@ -98,9 +107,15 @@ def write(outfile, *inputs, snaplen = SNAPLEN, basetime=None):
                        flags |= WS_FLAGS_PHY_2M
                elif modulation == P2G4_MOD_BLE_CODED:
                        flags |= WS_FLAGS_PHY_CODED
+                       if (int(row['packet_size'], 10) != 1): # If this is a FEC1, it has only the CI byte
+                               continue
 
-                       # The next row contains the FEC2 content:
+                       # The next row likely contains this packet FEC2:
                        row = rows[min_idx]
+                       if not is_coded_FEC2(row):
+                               # Otherwise the coded phy packet was aborted before the FEC2, so we just ignore it
+                               continue
+
                        rows[min_idx] = next(inputs[min_idx], None)
 
                        (orig_len, pdu_crc_FEC2) = get_data_and_len(row)
```

Commit:
fix: considers time order and FEC1 w/o FEC2

This modification ensures that packets are placed in chronological order
in the output pcap file. In addition, it is possible that some CODED PHY
packets are stopped between the FEC1 and FEC2. This modification takes
this into account.

Signed-off-by: Louka Yerly <louka.yerly@hes-so.ch>

Louka-Yerly commented:
@aescolar, I've included the requested changes in a new commit. I've been able to test it on my own and this new version is functional.

aescolar commented:
Thanks @Louka-Yerly the change is merged
